util/java/libtiled-java/src/main/java/org/mapeditor/io/TMXMapReader.java

```diff
@@ -83,18 +85,28 @@
     public static final long FLIPPED_VERTICALLY_FLAG = 0xFFFFFFFF40000000L;
     public static final long FLIPPED_DIAGONALLY_FLAG = 0xFFFFFFFF20000000L;
 
-    public static final long ALL_FLAGS = FLIPPED_HORIZONTALLY_FLAG
-            | FLIPPED_VERTICALLY_FLAG
-            | FLIPPED_DIAGONALLY_FLAG;
+    public static final long ALL_FLAGS =
+        FLIPPED_HORIZONTALLY_FLAG | FLIPPED_VERTICALLY_FLAG | FLIPPED_DIAGONALLY_FLAG;
+
+    public final TMXMapReaderSettings settings = new TMXMapReaderSettings();
 
     private Map map;
     private URL xmlPath;
     private String error;
     private final EntityResolver entityResolver = new MapEntityResolver();
     private TreeMap<Integer, TileSet> tilesetPerFirstGid;
-    public final TMXMapReaderSettings settings = new TMXMapReaderSettings();
-    private final java.util.Map<String, TileSet> cachedTilesets = new HashMap<>();
-    private final java.util.Map<Class<?>, Unmarshaller> cachedUnmarshallers = new HashMap<>();
+
+    /**
+     * Map of cached tilesets used when {@link TMXMapReaderSettings#reuseCachedTilesets} option is on
+     * TODO: case when multiple tilesets have same name but different sources
+     */
+    private java.util.Map<String, TileSet> cachedTilesets;
+
+    /**
+     * Keeping static map of unmarshaller pools allows to significantly increase map reading speed
+     * when reading maps in multiple threads
+     */
+    private static final java.util.Map<Class<?>, UnmarshallerPool> cachedUnmarshallers = new ConcurrentHashMap<>();
```

Comment:
I've been considering whether we could possibly get rid of the `cachedUnmarshallers` hash map, since it is only needed because we create a new `Unmarshaller` for each supported class. However, it should be possible to pass all used types to `JAXBContext.newInstance` and create only a single `Unmarshaller` supporting all types. If we do that we'd only need a single `UnmarshallerPool` in the new code. Would you be interested in trying this out, possibly with an eye on performance as well?

Reply:
yeah, great idea